MobileCenter/MobileCenterTests/MSCustomPropertiesTests.m

```diff
@@ -0,0 +1,194 @@
+#import <OCHamcrestIOS/OCHamcrestIOS.h>
+#import <XCTest/XCTest.h>
+
+#import "MSCustomProperties.h"
+#import "MSCustomPropertiesPrivate.h"
+
+@interface MSCustomPropertiesTests : XCTestCase
+
+@end
+
+@implementation MSCustomPropertiesTests
+
+- (void)testKeyValidate {
+  
+  // If
+  NSString *string = @"test";
+  NSDate *date = [NSDate dateWithTimeIntervalSince1970:0];
+  NSNumber *number = @0;
+  BOOL boolean = NO;
+  
+  // When
+  MSCustomProperties *customProperties = [MSCustomProperties new];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Null key.
+  // When
+  NSString *nullKey = nil;
+  [customProperties setString:string forKey:nullKey];
+  [customProperties setDate:date forKey:nullKey];
+  [customProperties setNumber:number forKey:nullKey];
+  [customProperties setBool:boolean forKey:nullKey];
+  [customProperties clearPropertyForKey:nullKey];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Invalid key.
+  // When
+  NSString *invalidKey = @"!";
+  [customProperties setString:string forKey:invalidKey];
+  [customProperties setDate:date forKey:invalidKey];
+  [customProperties setNumber:number forKey:invalidKey];
+  [customProperties setBool:boolean forKey:invalidKey];
+  [customProperties clearPropertyForKey:invalidKey];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Normal keys.
+  // When
+  [customProperties setString:string forKey:@"t1"];
+  [customProperties setDate:date forKey:@"t2"];
+  [customProperties setNumber:number forKey:@"t3"];
+  [customProperties setBool:boolean forKey:@"t4"];
+  [customProperties clearPropertyForKey:@"t5"];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(5));
+  
+  // Already contains keys.
+  // When
+  [customProperties setString:string forKey:@"t1"];
+  [customProperties setDate:date forKey:@"t2"];
+  [customProperties setNumber:number forKey:@"t3"];
+  [customProperties setBool:boolean forKey:@"t4"];
+  [customProperties clearPropertyForKey:@"t5"];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(5));
+}
+
+- (void)testSetString {
+  
+  // If
+  NSString *key = @"test";
+  
+  // When
+  MSCustomProperties *customProperties = [MSCustomProperties new];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Null value.
+  // When
+  NSString *nullValue = nil;
+  [customProperties setString:nullValue forKey:key];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(0));
+  
+  // Normal value.
+  // When
+  NSString *normalValue = @"test";
+  [customProperties setString:normalValue forKey:key];
+  
+  // Then
+  assertThat([customProperties properties], hasCountOf(1));
+}
```

Comment:
There is the special case of the empty string which is allowed as a value that you may want to add.